### PR TITLE
usersession/userd: support Cloudflare WARP URL scheme

### PR DESCRIPTION
**Pull request by schultetwin1**
Allow Cloudflare WARP URL scheme to be opened by xdg-open with in snap
to support WARP's user authorization work flow.

### CODE REVIEW

**Review by mardy — APPROVED** [review]
LGTM, but let's wait for some other review as well. :-)

File: usersession/userd/launcher.go
Comment (MiguelPires) [inline usersession/userd/launcher.go:83]: The static checks are failing because of this typo
```suggestion
		// com.cloudflare.warp: the scheme allows for specifying a package for
```

**Review by MiguelPires — COMMENTED** [review]
Looks good, but the build failed due to a typo

File: usersession/userd/launcher.go
Comment (schultetwin1) [reply usersession/userd/launcher.go]: Whoops, thank you! That's what you get for last minute comment changes.

**Review by MiguelPires — APPROVED** [review]
LGTM, thanks!

File: usersession/userd/launcher_test.go
Comment (bboozzoo) [inline usersession/userd/launcher_test.go:77]: can you add this at the end of the list?

Comment (schultetwin1) [reply usersession/userd/launcher_test.go]: For sure! I thought it was alphabetical but it looks like its not. Done!

File: usersession/userd/launcher.go
Comment (bboozzoo) [reply usersession/userd/launcher.go:86]: This still needs an update

**Comment by bboozzoo** [comment]
Also, a question which snap is affected and how come it's not being handled by xdg portal at this point? We expect all desktop setups to actually provide a working xdg-desktop-portal, while the code modified in this PR is really just a fallback.

**Comment by bboozzoo** [comment]
For reference, I've opened https://github.com/snapcore/snapd/pull/11200 which if lands would allow the scheme to be used without any changes on the snapd side provided there is a handler registered for it locally (as in an app with a desktop file that declares that scheme)

File: usersession/userd/launcher.go
Comment (schultetwin1) [reply usersession/userd/launcher.go]: Whoops, sorry about that. Moved now!

**Comment by schultetwin1** [comment]
> Also, a question which snap is affected and how come it's not being handled by xdg portal at this point? We expect all desktop setups to actually provide a working xdg-desktop-portal, while the code modified in this PR is really just a fallback.

We would like Chrome and Firefox (when deployed via snap) to prompt the user to open up the Cloudflare Warp desktop client when one of our web pages redirects to com.cloudflare.warp://{custom info here}.

So to answer your question, the affected snap would be Chrome or Firefox. From searching through Chromium and Firefox source code it looks like both still use `xdg-open` to open custom schemes in URIs (but I maybe wrong about that)

**Comment by schultetwin1** [comment]
> For reference, I've opened #11200 which if lands would allow the scheme to be used without any changes on the snapd side provided there is a handler registered for it locally (as in an app with a desktop file that declares that scheme)

That looks great! If that lands it will solve our problem as well. How do you want to proceed? My vote would be to merge this PR just in case it takes #11200 a long time to land

**Comment by bboozzoo** [comment]
> > For reference, I've opened #11200 which if lands would allow the scheme to be used without any changes on the snapd side provided there is a handler registered for it locally (as in an app with a desktop file that declares that scheme)
> 
> That looks great! If that lands it will solve our problem as well. How do you want to proceed? My vote would be to merge this PR just in case it takes #11200 a long time to land

I'm still waiting for the security team to take a look. Depending on the outcome of that we either land #11200 or this branch.

**Comment by schultetwin1** [comment]
> I'm still waiting for the security team to take a look. Depending on the outcome of that we either land #11200 or this branch.

Do we need to block this on the outcome of #11200? If #11200 is merged then this code will be a no-op anyways and if it's not merged then this code should be merged.

**Comment by alexmurray** [comment]
With my security team hat on, I would prefer we try and land #11200 as this is a much better generic solution than the current hard-coded list of URLs

**Comment by schultetwin1** [comment]
That makes sense to me! @bboozzoo do you have a general sense of when you hope to merge #11200? In the next couple days, weeks, or months?

I realize you may not have an exact time but a generally idea would be helpful for helping us communicate to our customers. Thanks!

**Comment by bboozzoo** [comment]
> That makes sense to me! @bboozzoo do you have a general sense of when you hope to merge #11200? In the next couple days, weeks, or months?
> 
> I realize you may not have an exact time but a generally idea would be helpful for helping us communicate to our customers. Thanks!

Pushed a little tweak there and I hope it will land today once the CI job is finished.

**Comment by bboozzoo** [comment]
#11200 has been merged.

**Comment by schultetwin1** [comment]
Perfect! Thank you!

**Pull request closed without merging**
